File: 2018/5/puzzle5.erl
-module(puzzle5).
-export([main/0]).
-compile([export_all]).

main() ->
    {ok, Bin} = file:read_file("input.txt"),
    Input = string:trim(binary_to_list(Bin)),
    {{part1, start1(Input)},
     {part2, start2(Input)}}.

start1(L) ->
    length(react(L)).

start2(L) ->
    lists:min([start1([C || C <- L, C =/= X, C =/= X - 32]) || 
                  X <- lists:seq($a, $z)]).

react(L) ->
    react(L, []).
             
react([], L) ->
    L;
react([C1|L1], [C2|L2]) when abs(C1 - C2) == 32 ->
    react(L1, L2);
react([C|L1], L2) ->
    react(L1, [C|L2]).

       

File: 2018/5/src/Puzzle5.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

public class Puzzle5 {
    public static void main(String[] args)
            throws FileNotFoundException, IOException {
        try (BufferedReader reader = new BufferedReader(
                new FileReader("input.txt"))) {
            String orig = reader.readLine();
            char[] buf = orig.toCharArray();

            long t0 = System.nanoTime();
            int len = react(buf);
            int part1 = len;

            int minlen = Integer.MAX_VALUE;
            StringBuilder s = new StringBuilder();
            for (char c = 'a'; c != 'z'; c++) {
                s.setLength(0);
                buf = orig.toCharArray();

                for (int i = 0; i < buf.length; i++) {
                    if (buf[i] != c && buf[i] != c - 32) {
                        s.append(buf[i]);
                    }
                }

                buf = s.toString().toCharArray();
                len = react(buf);
                minlen = Math.min(len, minlen);
            }
            long elapsed = System.nanoTime() - t0;
            int part2 = minlen;

            System.out.format("Part 1 = %d, Part 2 = %d, msecs = %d%n", part1,
                    part2, TimeUnit.NANOSECONDS.toMillis(elapsed));
        }
    }

    static private int react(char[] buf) {
        int l = buf.length;

        while (true) {
            int l0 = react_once(buf, l);

            if (l0 == l)
                return l0;
            else
                l = l0;
        }
    }

    static private int react_once(char[] buf, int len) {

        int src = 0;
        int tgt = 0;

        while (true) {
            if (src == len)
                break;

            if (src + 1 == len) {
                buf[tgt++] = buf[src++];
                break;
            }

            char x = buf[src];
            char y = buf[src + 1];

            if (Math.abs(x - y) == 32) {
                src += 2;
            } else {
                buf[tgt++] = buf[src++];
            }
        }

        return tgt;
    }
}
